components/merkledb/src/proof_list_index/proof.rs

```diff
@@ -12,302 +12,603 @@
+    /// Otherwise, an error is returned.
     pub fn validate(
         &self,
         expected_list_hash: Hash,
         len: u64,
-    ) -> Result<Vec<(u64, &V)>, ListProofError> {
-        let mut vec = Vec::new();
-        let height = len.next_power_of_two().trailing_zeros() as u8 + 1;
-        let root_hash = self.collect(ProofListKey::new(height, 0), &mut vec)?;
-
-        match *self {
-            ListProof::Absent(_) => {
-                if expected_list_hash != root_hash {
-                    return Err(ListProofError::UnmatchedRootHash);
-                }
-            }
-            _ => {
-                if HashTag::hash_list_node(len, root_hash) != expected_list_hash {
-                    return Err(ListProofError::UnmatchedRootHash);
-                }
-            }
+    ) -> Result<&[(u64, V)], ListProofError> {
+        let tree_root = self.collect(len)?;
+        if HashTag::hash_list_node(len, tree_root) == expected_list_hash {
+            Ok(&self.values)
+        } else {
+            Err(ListProofError::UnmatchedRootHash)
         }
-        Ok(vec)
     }
 }
 
-impl<V: Serialize> Serialize for ListProof<V> {
-    fn serialize<S>(&self, ser: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        use self::ListProof::*;
-        let mut state;
-        match *self {
-            Full(ref left_proof, ref right_proof) => {
-                state = ser.serialize_struct("Full", 2)?;
-                state.serialize_field("left", left_proof)?;
-                state.serialize_field("right", right_proof)?;
-            }
-            Left(ref left_proof, ref option_hash) => {
-                if let Some(ref hash) = *option_hash {
-                    state = ser.serialize_struct("Left", 2)?;
-                    state.serialize_field("left", left_proof)?;
-                    state.serialize_field("right", &hash)?;
-                } else {
-                    state = ser.serialize_struct("Left", 1)?;
-                    state.serialize_field("left", left_proof)?;
-                }
-            }
-            Right(ref hash, ref right_proof) => {
-                state = ser.serialize_struct("Right", 2)?;
-                state.serialize_field("left", &hash)?;
-                state.serialize_field("right", right_proof)?;
-            }
-            Leaf(ref val) => {
-                state = ser.serialize_struct("Leaf", 1)?;
-                state.serialize_field("val", val)?;
-            }
-            ListProof::Absent(ref proof) => {
-                state = ser.serialize_struct("Absent", 2)?;
-                state.serialize_field("length", &proof.length)?;
-                state.serialize_field("hash", &proof.merkle_root)?;
-            }
-        }
-        state.end()
-    }
+/// An error that is returned when the list proof is invalid.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Fail)]
+pub enum ListProofError {
```

Comment:
Just a bit of thoughts: currently ListProofError can be splitted into two subsets -- errors that user do care about and other errors. The first subset contains only one type of errors: `UnmatchedRootHash`. This error means that proof (or the root hash) was incorrect. All the rest ones can be named `MalformedProofError` and user can do nothing about it (it will be wierd to attempt fix errors in the received proof). Receiving that kind of error would mean that something is wrong on either side that created proof or in the communication channel (e.g. due to MITM attack JSON representation of the proof was intentionally corrupted).
So won't it make sence to have only two types of error: UnmatchedRootHash and MalformedProof(details)? I think it will simplify the user code.

Reply:
If we go with `MapProof` API, we could have a separate data structure for a checked list proof, which would contain list hash and checked elements. (I'm not quite fond of the specifics of `CheckedMapProof`, though; namely, the fact that it consumes the `MapProof` instead of holding references to elements. I vaguely remember that there were Java binding considerations involved in this decision, but I may be mistaken.) In this case, comparing hash of `CheckedListProof` with the reference hash could be done by the caller, and the `validate` method would yield only "malformed proof" errors.